third_party/amd/python/triton_amd.cc

```diff
@@ -114,4 +114,14 @@ void init_triton_amd(py::module &&m) {
     if (auto *openclVersion = module->getNamedMetadata("opencl.ocl.version"))
       module->eraseNamedMetadata(openclVersion);
   });
+
+  m.def("set_fn_arg_inreg", [](llvm::Function *fn) {
+    for (unsigned I = 0; I < fn->arg_size(); ++I) {
+      llvm::Argument &Arg = *fn->getArg(I);
+      // Check for incompatible attributes.
+      if (Arg.hasByRefAttr() || Arg.hasNestAttr())
+        break;
+      Arg.addAttr(llvm::Attribute::InReg);
```

Comment:
FYI, in IREE we found that capping the number of `InReg` args between 8 and 16 gives best performance on our kernels.

Reply:
I learned that the backend compiler will decide how many args will actually be pre-set into sgprs. It's intuitive that we want as many as possible args to be in sgprs. But what is the benefit of limiting the number of args with InReg attr?

Reply:
I tried different limits and, empirically, it hurts performance to preload more than 8, at least in our benchmarks. I dumped the number of kernargs that are actually preloaded and saw that there are a few with >= 8, the maximum is 10. This was a small difference (~0.2% of total runtime) that consistently showed across 1000 benchmark runs.

Reply:
This is interesting. I'd like to benchmark triton kernels to see if this is also the case.
> I dumped the number of kernargs that are actually preloaded

Can you clarify how did you do this? Dump the ISA and count the `s_load` instructions?

Reply:
I just added some prints in the analogous code to this one: how many arguments there were, how many were rejected by this if statement above. Just to get a feel of what the distribution of preloaded/normal kernel args was in my benchmark.

Reply:
Got it. Thanks for the clarification. So basically you got the distribution of kernel args that are "tagged" with this `InReg` attribute. I learned that this `InReg` attribute is a hint, so "being tagged with InReg" can be different from "being preloaded into regs". 
I just did some experiment with gemm 384x384x384, which has 9 kernel args in total. I did see some reversely linear relationship between number of args tagged and number of dword s_loaded in the isa like this 

#InReg | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
#s_load_dword | 11 | 9 | 7 | 5 | 5 | 4 | 3 | 2 | 1 | 0

The more interesting part is the perf. For #InReg between 0 to 8, the perf is about 14 ~ 17 us. However, when #InReg=9, i.e. all args are tagged, perf becomes ~ 5us. And in this particular case, none of the args has incompatible attributes.
I'll talk to compiler/firmware folks internally to learn more about how it works.

Reply:
Thanks for the comment @kuhar and thanks for the due diligence @zhanglx13! I think the above makes intuitive sense in the manner that if you can fit all parameters into registers you don't need to wait for anything so there is a drop in latency. While if you have some argument to load for memory you are bound to wait at least some minimal latency. I feel it's complicated w.r.t. other factors the compiler decide; so likely there will be no one solution for all and even "good default number". It might be workload dependent (@kuhar was talkinga bout another whole model workload where we have lots of kernels and @zhanglx13 is talking specificially about a particular kernel). This might be a tunable thing eventually.

For now I think marking all arguments as preferred to load into registers is a good first step. We can proceed from here.